### PR TITLE
feat(api): Add REST API service for document validation and diffing

### LINKED ISSUE #9: [Phase 3] Task 3.3: REST API Service
## Description
Create a REST API service using FastAPI for document validation and diffing, designed for Railway deployment.

## Dependencies
- Task 3.1: Python Library API

## Acceptance Criteria
- `POST /api/v1/validate` endpoint validates a document
- `POST /api/v1/diff` endpoint compares two documents
- `GET /health` endpoint for health checks (required for Railway)
- API accepts YAML in request body
- API returns JSON responses
- API includes OpenAPI/Swagger documentation
- API handles errors with appropriate HTTP status codes
- API supports CORS for cross-origin requests
- API includes request/response examples
- API reads port from `PORT` environment variable (Railway requirement)
- API binds to `0.0.0.0` for Railway deployment
- Logging configured for production environment

## Tests
- Unit test: Validate endpoint with valid document
- Unit test: Validate endpoint with invalid document
- Unit test: Diff endpoint compares documents
- Unit test: Health check endpoint returns 200
- Unit test: API returns correct HTTP status codes
- Unit test: API handles malformed requests
- Unit test: API reads PORT from environment variable
- Integration test: API endpoints work end-to-end
- Acceptance test: API responses match OpenAPI spec

## Files to Create
- `src/yaml_diffs/api_server/__init__.py`
- `src/yaml_diffs/api_server/main.py`
- `src/yaml_diffs/api_server/routers/validate.py`
- `src/yaml_diffs/api_server/routers/diff.py`
- `src/yaml_diffs/api_server/routers/health.py`
- `src/yaml_diffs/api_server/schemas.py` (API request/response models)
- `src/yaml_diffs/api_server/config.py` (environment configuration)
- `tests/test_api_server.py`
- `docs/api_server.md`

## Phase
Phase 3: Interface Layers

### PR DESCRIPTION
**Pull request by noamoss**
## Summary

This PR implements a complete REST API service using FastAPI for document validation and diffing, designed for Railway deployment. The API exposes the existing library functionality via HTTP endpoints with proper error handling, CORS support, and OpenAPI documentation.

## Related Issues

Fixes #9

## Changes Made

- **API Server Implementation:**
  - Created FastAPI application with three main endpoints:
    - `POST /api/v1/validate` - Validates YAML documents against OpenSpec schema and Pydantic models
    - `POST /api/v1/diff` - Compares two YAML documents and returns detected changes
    - `GET /health` - Health check endpoint for Railway monitoring
  - Added root endpoint (`GET /`) with API information

- **Configuration:**
  - Environment variable handling (PORT, CORS, logging)
  - Railway-ready configuration (reads PORT env var, binds to 0.0.0.0)
  - Configurable CORS settings

- **Error Handling:**
  - Comprehensive exception handlers for all custom exceptions
  - Proper HTTP status codes (400, 422, 500)
  - Consistent error response format

- **Request/Response Models:**
  - Pydantic schemas for all endpoints
  - Request validation
  - Response serialization

- **Documentation:**
  - Complete API server documentation (`docs/api_server.md`)
  - Railway deployment guide
  - Request/response examples
  - Environment variables documentation

- **Testing:**
  - Comprehensive test suite (26 tests)
  - 83% code coverage
  - Tests for all endpoints, error cases, and edge cases
  - Integration tests for end-to-end workflows

- **Dependencies:**
  - Added `httpx>=0.24.0` to dev dependencies for testing

## Testing

- [x] All tests pass (26/26)
- [x] Pre-commit hooks pass
- [x] Code coverage: 83%
- [x] Linting: All checks pass
- [x] Type checking: All checks pass
- [x] OpenAPI spec generated successfully

## Checklist

- [x] Tests added/updated
- [x] Documentation updated
- [x] All acceptance criteria met
- [x] Railway deployment ready

## Acceptance Criteria Status

All acceptance criteria from issue #9 have been met:

- [x] `POST /api/v1/validate` endpoint validates a document
- [x] `POST /api/v1/diff` endpoint compares two documents
- [x] `GET /health` endpoint for health checks (required for Railway)
- [x] API accepts YAML in request body
- [x] API returns JSON responses
- [x] API includes OpenAPI/Swagger documentation
- [x] API handles errors with appropriate HTTP status codes
- [x] API supports CORS for cross-origin requests
- [x] API includes request/response examples
- [x] API reads port from `PORT` environment variable (Railway requirement)
- [x] API binds to `0.0.0.0` for Railway deployment
- [x] Logging configured for production environment

## Files Created

- `src/yaml_diffs/api_server/__init__.py`
- `src/yaml_diffs/api_server/main.py`
- `src/yaml_diffs/api_server/config.py`
- `src/yaml_diffs/api_server/schemas.py`
- `src/yaml_diffs/api_server/routers/__init__.py`
- `src/yaml_diffs/api_server/routers/health.py`
- `src/yaml_diffs/api_server/routers/validate.py`
- `src/yaml_diffs/api_server/routers/diff.py`
- `tests/test_api_server.py`
- `docs/api_server.md`

## Files Modified

- `pyproject.toml` - Added httpx to dev dependencies

## Next Steps

- Railway deployment configuration (Issue #14)
- MCP server implementation (Issue #37)

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/yaml_diffs/api_server/main.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    )
+
+
+# Include routers
+app.include_router(health.router)
+app.include_router(validate.router)
+app.include_router(diff.router)
+
+
+@app.on_event("startup")
```
Comment (noamoss) [inline src/yaml_diffs/api_server/main.py]: **Issue: Deprecated FastAPI Pattern**

The `@app.on_event("startup")` and `@app.on_event("shutdown")` decorators are deprecated in FastAPI 0.95+ in favor of the lifespan context manager pattern.

**Problem:**
- FastAPI 0.100.0+ (as specified in dependencies) deprecates `on_event` decorators
- This pattern will be removed in future FastAPI versions

**Suggested Fix:**
Use the lifespan context manager pattern instead:

```python
from contextlib import asynccontextmanager
from typing import AsyncGenerator

@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    """Lifespan context manager for startup and shutdown events."""
    # Startup
    logger.info(f"Starting {settings.app_name} v{__version__}")
    logger.info(f"Server will bind to {settings.host}:{settings.port_from_env}")
    logger.info(f"Log level: {settings.log_level}")
    yield
    # Shutdown
    logger.info(f"Shutting down {settings.app_name}")

# Create FastAPI app with lifespan
app = FastAPI(
    title=settings.app_name,
    description="REST API for validating and diffing Hebrew legal documents in YAML format",
    version=__version__,
    docs_url="/docs",
    redoc_url="/redoc",
    openapi_url="/openapi.json",
    lifespan=lifespan,  # Add lifespan parameter
)
```

Then remove the `@app.on_event` decorators and their functions.

**Why this matters:**
- Future-proofs the code against FastAPI updates
- Follows current FastAPI best practices
- More explicit and testable lifecycle management

File: src/yaml_diffs/api_server/main.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    CORSMiddleware,
+    allow_origins=settings.cors_origins,
+    allow_credentials=settings.cors_allow_credentials,
+    allow_methods=settings.cors_allow_methods,
+    allow_headers=settings.cors_allow_headers,
+)
+
+
+# Exception handlers
+@app.exception_handler(YAMLLoadError)
```
Comment (noamoss) [inline src/yaml_diffs/api_server/main.py]: **Issue: Error Response Format Inconsistency**

There's an inconsistency between how routers raise errors (`HTTPException` with `detail`) and how global exception handlers format responses (`JSONResponse` with `content`).

**Problem:**
- Routers (validate.py, diff.py) raise `HTTPException` with `detail` field
- Global exception handlers return `JSONResponse` with `content` field
- Both end up in the response, but the format differs

**Current behavior:**
- Router raises: `HTTPException(detail={"error": "...", "message": "..."})`
- Global handler returns: `JSONResponse(content={"error": "...", "message": "..."})`
- FastAPI wraps `HTTPException.detail` in `{"detail": {...}}` format

**Suggested Fix:**
Since global exception handlers exist, routers should just let exceptions propagate. However, if keeping router-level handling, ensure consistency:

1. **Option 1 (Recommended):** Remove try-except from routers, let global handlers catch everything
2. **Option 2:** If keeping router handling, use consistent format:

```python
# In routers, raise HTTPException with detail that matches global handler format
raise HTTPException(
    status_code=status.HTTP_400_BAD_REQUEST,
    detail={
        "error": "YAMLLoadError",
        "message": str(e),
        "details": [{"file_path": e.file_path}] if e.file_path else None,
    },
)
```

Then FastAPI will wrap it as `{"detail": {...}}` which matches the global handler pattern.

**Why this matters:**
- Consistent API error responses
- Easier for API consumers to handle errors
- Better developer experience

**Comment by noamoss** [comment]
## 📋 Review Feedback Addressed

I've applied the following changes based on code review:

### ✅ Changes Made:

1. **Removed duplicate exception handling from routers** ([comment #2557836773](https://github.com/noamoss/yaml_diffs/pull/38#discussion_r2557836773))
   - Removed try-except blocks from `validate.py` and `diff.py` routers
   - Let global exception handlers in `main.py` catch all exceptions
   - Reduces code duplication and follows DRY principle

2. **Migrated to FastAPI lifespan pattern** ([comment #2557837012](https://github.com/noamoss/yaml_diffs/pull/38#discussion_r2557837012))
   - Replaced deprecated `@app.on_event` decorators with `lifespan` context manager
   - Future-proofs the code against FastAPI updates
   - Follows current FastAPI best practices

3. **Made CORS more restrictive by default** ([comment #2557837169](https://github.com/noamoss/yaml_diffs/pull/38#discussion_r2557837169))
   - Changed default CORS origins from `["*"]` to `[]` (empty list)
   - Requires explicit configuration via `CORS_ORIGINS` environment variable
   - Added warning if CORS is set to `*` in production environment
   - Follows security best practices (principle of least privilege)

4. **Added maximum length validation to YAML input fields** ([comment #2557837459](https://github.com/noamoss/yaml_diffs/pull/38#discussion_r2557837459))
   - Added `max_length=10_000_000` (10MB) to all YAML string fields
   - Configurable via `MAX_YAML_SIZE` environment variable
   - Prevents memory exhaustion and DoS attacks

5. **Moved logging configuration to lifespan startup event** ([comment #2557837948](https://github.com/noamoss/yaml_diffs/pull/38#discussion_r2557837948))
   - Logging now configured in lifespan startup instead of module import time
   - Ensures logging is configured at the right time in application lifecycle
   - Better for testing (can configure logging per test)

6. **Updated tests to match new error response format**
   - Tests now expect error responses directly from JSONResponse (not wrapped in `detail`)
   - All 26 tests pass with 89% code coverage

### ✅ Validation:
- All 26 tests passed
- Code coverage: 89% (improved from 83%)
- Linting: All checks pass
- Changes committed locally

### 📦 Next Step:
Pushing changes to remote branch...

---
Ready for re-review! 🚀

File: src/yaml_diffs/api_server/main.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    """Lifespan context manager for startup and shutdown events.
+
+    Configures logging and logs startup/shutdown information.
+    """
+    # Startup
+    configure_logging(settings.log_level)
+    logger.info(f"Starting {settings.app_name} v{__version__}")
+    logger.info(f"Server will bind to {settings.host}:{settings.port_from_env}")
+    logger.info(f"Log level: {settings.log_level}")
+    yield
```
Comment (noamoss) [inline src/yaml_diffs/api_server/main.py]: **Minor Issue: Missing `yield` Statement**

The lifespan context manager is missing the `yield` statement between startup and shutdown code. Currently, the shutdown code (line 43) will execute immediately after startup, not during actual shutdown.

**Current Code:**
```python
@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    # Startup
    configure_logging(settings.log_level)
    logger.info(f"Starting {settings.app_name} v{__version__}")
    logger.info(f"Server will bind to {settings.host}:{settings.port_from_env}")
    logger.info(f"Log level: {settings.log_level}")
    # Shutdown
    logger.info(f"Shutting down {settings.app_name}")
```

**Problem:**
The shutdown log will execute immediately after startup, not when the application actually shuts down. The `yield` statement is required to separate startup from shutdown.

**Suggested Fix:**
```python
@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    """Lifespan context manager for startup and shutdown events.

    Configures logging and logs startup/shutdown information.
    """
    # Startup
    configure_logging(settings.log_level)
    logger.info(f"Starting {settings.app_name} v{__version__}")
    logger.info(f"Server will bind to {settings.host}:{settings.port_from_env}")
    logger.info(f"Log level: {settings.log_level}")
    yield  # Application runs here
    # Shutdown
    logger.info(f"Shutting down {settings.app_name}")
```

**Why this matters:**
- Ensures shutdown code runs at the correct time
- Follows the proper async context manager pattern
- Shutdown logging will actually occur when the app stops

File: src/yaml_diffs/api_server/schemas.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+from yaml_diffs.models import Document
+
+
+def _get_max_yaml_size() -> int:
+    """Get maximum YAML size from environment variable.
+
+    Returns:
+        Maximum YAML size in bytes, defaulting to 10MB if not set or invalid.
+    """
+    import logging
```
Comment (noamoss) [inline src/yaml_diffs/api_server/schemas.py]: **Minor: Import Organization**

The `import logging` statement is inside the function (line 23), which is fine but could be moved to the top of the file for consistency with Python conventions.

**Current Code:**
```python
def _get_max_yaml_size() -> int:
    """Get maximum YAML size from environment variable."""
    import logging  # Import inside function
    
    value = os.getenv("MAX_YAML_SIZE", "10_000_000")
    ...
```

**Suggested Fix:**
Move the import to the top of the file with other imports:

```python
from __future__ import annotations

import logging  # Add here
import os
from typing import Any, Optional

from pydantic import BaseModel, Field

from yaml_diffs.diff_types import DocumentDiff
from yaml_diffs.models import Document


def _get_max_yaml_size() -> int:
    """Get maximum YAML size from environment variable."""
    value = os.getenv("MAX_YAML_SIZE", "10_000_000")
    try:
        return int(value)
    except ValueError:
        logging.warning(
            f"Invalid MAX_YAML_SIZE environment variable: {value}. Using default: 10_000_000"
        )
        return 10_000_000
```

**Why this matters:**
- Follows PEP 8 import conventions (imports at top of file)
- Makes dependencies more visible
- Consistent with rest of codebase
- Slightly better performance (import happens once, not on every function call, though this is negligible here since it's only called at module load time)
